### PR TITLE
Don't use float equality for AST comparisons

### PR DESCRIPTION
**Pull request by Clyybber**
## Summary
Float equality lacks the substitution nor reflexivity properties usually
expected from an equality operator, so it's not correct to use float
equality in AST comparisons. This PR changes it so that they are
compared for bit equality.

## Details
*   `0.0`  and  `-0.0`  are not being considered equal by the compiler
anymore, this affects:
    -  static parameters in generic types and procedures, see `tests/statictypes/tstatictypes.nim`
    - default arguments for forward declarations, see `tests/errmsgs/tforwarddecl_defaultparam.nim`
    - ```macros.`==`(a, b: NimNode)```, see `tests/lang_callable/macros/tmacros_various.nim`
    - term-rewriting macros, see `tests/lang_experimental/trmacros/trmacros_various2.nim`

*  `trees.exprStructuralEquivalentStrictSym`  and it's only usage in 
`sem/semfoldnim`  have been removed

### CODE REVIEW

File: compiler/ast/trees.nim [hunk 1]
```diff
@@ -40,12 +40,16 @@ proc cyclicTree*(n: PNode): bool =
...
   cyclicTreeAux(n, visited)
 
-proc sameFloatIgnoreNan(a, b: BiggestFloat): bool {.inline.} =
-  ## ignores NaN semantics, but ensures 0.0 == -0.0, see #13730
-  cast[uint64](a) == cast[uint64](b) or a == b
+template cmpFloatRep*(a, b: BiggestFloat): bool =
+  ## Compares the bit-representation of floats `a` and `b`
+  # Special handling for floats, so that floats that have the same
+  # value but different bit representations are treated as different constants
+  # As opposed to float equality this does not lack the substition or
```
Comment (zerbina) [inline compiler/ast/trees.nim]:
```suggestion
  # Compared to float equality, this does not lack the substitution and
```
A small typo fix and soft wording change suggestion.

File: compiler/sem/semfold.nim [hunk 2]
```diff
@@ -378,8 +379,16 @@ proc evalOp*(m: TMagic, n, a, b, c: PNode; idgen: IdGenerator; g: ModuleGraph):
...
-        exprStructuralEquivalentStrictSym(a, b))), n, idgen, g)
+    g.config.internalAssert(a.kind in {nkSym, nkNilLit} and
+                            b.kind in {nkSym, nkNilLit},
+                            n.info, "mEqProc: invalid AST")
+    let isEqual = if a.kind != b.kind:
+                    false
+                  elif a.kind == nkSym: # and b.kind == nkSym:
+                    a.sym == b.sym
+                  else: # a.kind == nkNilLit and b.kind == nkNilLit
+                    true
```
Comment (zerbina) [inline compiler/sem/semfold.nim]:
```suggestion
    let isEqual =
      if a.kind != b.kind:  false
      elif a.kind == nkSym: a.sym == b.sym # b.kind == nkSym
      else:                 true # a.kind == b.kind == nkNilLit
```
Soft suggestion regarding code style (recently introduced multi-line if/case/etc. expression tend to use this formatting).

File: tests/statictypes/tstatictypes.nim [hunk 3]
```diff
@@ -411,4 +415,11 @@ block coercion_to_static_type:
       result = 2.1
 
   # the call must be fully evaluated at compile-time
-  doAssert static[int](get()) == 1
\ No newline at end of file
+  doAssert static[int](get()) == 1
+
```
Comment (zerbina) [inline tests/statictypes/tstatictypes.nim:419]:
```suggestion
```
Very minor nitpick: superfluous whitespace.

**Review by zerbina — APPROVED** [review]
A nice fix of a floating-point related edge case, thank you! The review comments I left are very minor and only concern code style and wording of some comments.

For the PR message, could you note down what the affected language features are (e.g., static parameters for generic types and procedures, term-rewriting macros, etc.; can go into the details section), for easier future assessment of what changed?

**Comment by Clyybber** [comment]
/merge

Thanks for the review!
